### ISSUE TITLE
Can't update or install any packages without root permission

### ISSUE THREAD

**Issue by sedabull** [atom/atom#5492]
* Ubuntu 14.04
* Atom 0.178.0
* apm 0.136.0

It is impossible for me to update or install any packages from the atom ui. I get an error message like this:
![atom-update-error](https://cloud.githubusercontent.com/assets/9311295/6141355/2a450964-b172-11e4-9780-d67ed56741e5.png)

apm gives the same error on the command line, unless sudo is used. In which case it works fine. On a possibly related note, packages linked via `apm link --dev` show up in the installed packages view like they should, but are not activateable.

**Comment by gau1991** [comment]
Same here on Fedora

**Comment by izuzak** [comment]
This looks very similar to https://discuss.atom.io/t/solved-new-package-installs-updates-generates-errors/13152 -- can you check that discussion and see if the advice there helps?

**Comment by GeorgeSapkin** [comment]
Removing .atom directory didn't work for me on Fedora 21 x64.

**Comment by gau1991** [comment]
This discussion not worked for me...

**Comment by mehcode** [comment]
Something similar occurs for me on update that didn't occur on the last version.
![screenshot from 2015-02-11 06 28 21](https://cloud.githubusercontent.com/assets/753919/6148528/304b7ede-b1b7-11e4-8c7c-393338be7ddc.png)

**Comment by joefitzgerald** [comment]
See also: https://github.com/atom-community/autocomplete-plus/issues/260 @nathansobo @kevinsawicki @thedaniel @maxbrunsfeld - v0.178.0 seems to be broken - at least on Linux.

Works for me on OS X though:

![screen shot 2015-02-11 at 8 03 23 am](https://cloud.githubusercontent.com/assets/744740/6149233/7f3708b2-b1c4-11e4-81c4-65428fa8a1e4.png)

![screen shot 2015-02-11 at 8 02 31 am](https://cloud.githubusercontent.com/assets/744740/6149234/7f3927a0-b1c4-11e4-8db1-487239b7c08d.png)

![screen shot 2015-02-11 at 8 02 38 am](https://cloud.githubusercontent.com/assets/744740/6149235/7f39363c-b1c4-11e4-8040-31bbdd0c7363.png)

**Comment by edgard** [comment]
Same here, maybe it has something todo with pr #5385?

I tried setting ATOM_HOME but still doesn't work.

**Comment by mnquintana** [comment]
Same issue here on Ubuntu 14.04 after updating to Atom 0.178.0.

**Comment by olmokramer** [comment]
And the same on Debian 8

**Comment by kevinsawicki** [comment]
What are the folder permissions on `~/.atom/.apm`?

**Comment by olmokramer** [comment]
No such file/directory exists in my `~/.atom` directory

**Comment by kevinsawicki** [comment]
What are the folder permissions on `~/.atom`?

**Comment by olmokramer** [comment]
`~/.atom` has permissions `755`

**Comment by Bewahn** [comment]
I've got a very similar output as mehcode when trying to update linter-clang. So apm tries to write to /var/lib/jenkins which should not happen when installing a package as user, I guess...

**Comment by kevinsawicki** [comment]
@Bewahn what is your `HOME` environment variable set to?

**Comment by Bewahn** [comment]
It's set to my home directory. $HOME/atom is writable for the user.

**Comment by mehcode** [comment]
@kevinsawicki The weird thing here is look at the EACCESS path. `/var/lib/jenkins` --- I've never installed jenkins, I don't think atom uses jenkins, and it shouldn't be trying paths inside of `/var/lib` to begin with.

**Comment by kevinsawicki** [comment]
> The weird thing here is look at the EACCESS path. /var/lib/jenkins

Yeah, apologies for this, this was because of a file generated on the CI machines ending up in the distribution.

Just fixed this over in https://github.com/atom/apm/commit/230f9d627594f58a600d85167c686aaea59287fa

**Comment by mehcode** [comment]
@kevinsawicki It's fine. Just a confusing error from my end and I'm glad it got resolved quickly. Thank you.

**Comment by edgard** [comment]
[edgard@lynxarc ~]$ cat /usr/share/atom/resources/app/apm/.apmrc
cache = /.atom/.apm

[edgard@lynxarc ~]$ rpm -ql atom | grep -i apmrc
/usr/share/atom/resources/app/apm/.apmrc

It seems like the .apmrc in the package is being generated with a wrong path for the the cache.

**Comment by kevinsawicki** [comment]
@edgard can you check the file permissions/owner on `/usr/share/atom/resources/app/apm/.apmrc`?

**Comment by edgard** [comment]
Sure, there you go:

-rw-r--r-- 1 root root 20 Feb 10 20:26 /usr/share/atom/resources/app/apm/.apmrc

**Comment by kevinsawicki** [comment]
0.179 is now available, can people try upgrading and see if this issue is now fixed?

https://github.com/atom/atom/releases/tag/v0.179.0

**Comment by mnquintana** [comment]
0.179.0 resolved it for me!

**Comment by olmokramer** [comment]
Yes it works for me! Thanks!

**Comment by sedabull** [comment]
OP here, Upgrading to 0.179.0 fixed the update and installation problem, but I still can't get my locally linked package to activate. I've tried both `apm link --dev` and the normal `apm link`, but neither works. I can download my package from atom.io and activate the old version, but this bug makes it impossible to test local changes. The weird thing is I can see it in the installed packages view, and even play with the settings, but the activation command simply isn't available via keybinding or the command pallete.

**Comment by kevinsawicki** [comment]
@sedabull when you have your package linked, are you seeing any errors in the dev tools console in Atom after the window loads?

**Comment by kevinsawicki** [comment]
@sedabull also, can you link to the package you are having issues linking?

**Comment by kevinsawicki** [comment]
I'm guessing it is quantum-shell (from the screenshot).

I just ran `apm develop quantum-shell` and was able to load it in a dev mode window and see the toggle command in the command palette.

I also logged in `activate()` and saw the message go to the console when the toggle command was run.

Another thing to try from the dev tools is run `atom.packages.resolvePackagePath('quantum-shell')` to see where Atom thinks the package is.

**Comment by sedabull** [comment]
@kevinsawicki, I am not seeing any errors in the dev tools. When I do an `apm link --dev` I see this:
![dev-link-error](https://cloud.githubusercontent.com/assets/9311295/6160332/a1e693ee-b22f-11e4-9a72-5d8f404664e8.png)
It prompts me to install the package, and the toggle command is unavailable. Then when I try to get the resolved path:
![resolve-package-path](https://cloud.githubusercontent.com/assets/9311295/6160352/ba64d084-b22f-11e4-8176-e20d4466a908.png)
I used `web-browser` as a control to show that it does work. This is what happens on a normal `apm link`:
![normal-link](https://cloud.githubusercontent.com/assets/9311295/6160367/d9571efc-b22f-11e4-9075-3f75e8b9351a.png)
The toggle command is available in the command pallete, but it doesn't work. Not only that, after dispatching the command once, it disappears from the command pallete, with no output to the dev tools.

**Comment by sedabull** [comment]
Also, doing an `apm link --dev` and then `atom --dev` has the same behavior as a normal `apm link` and `atom`. My package shows up in the development package view (without a prompt to install), and the toggle command is available, but doesn't work and disappears after a single dispatch without any output to the dev tools.

**Comment by kevinsawicki** [comment]
@sedabull did the issue that you are seeing with a linked package only start after upgrading to 0.178?

**Comment by sedabull** [comment]
yes, I've been working exclusively on `quantum-shell` for the past week using 0.177.0, and this problem didn't present itself until immediately after I upgraded.

**Comment by sedabull** [comment]
@kevinsawicki, I know you're probably really busy, but I was just wondering if there are any updates on this issue, and/or if there's any more troubleshooting I can do from my end. I've tried pretty much every permutation I can think of `apm link`, `apm link --dev`, `atom`, `atom --dev`, I've even tried hard copying my local project into `~/.atom/packages` and `~/.atom/dev/packages`, but nothing seems to be working.

**Comment by sedabull** [comment]
I have a few updates, here are reproduction steps:
* in local repo, run `apm link --dev`
* in local repo, run `atom --dev`
* in dev tools, run `atom.packages.resolvePackagePath('quantum-shell');`, returns `/path/to/local/repo`
* in dev tools, run `atom.packages.activatePackage('quantum-shell');`, returns some kind of object
* in dev tools, run `atom.packages.isPackageActive('quantum-shell');`, returns `false`
* in dev tools, run `atom.commands.dispatch(atom.views.getView(atom.workspace), 'quantum-shell:toggle')`, returns `false` and as a side effect removes the `Quantum Shell: Toggle` option from the command pallete

I'm unsure of what could cause my package not to activate, but hopefully these steps will help you get to the bottom of it, please let me know if there's anything else I can do, I'm happy to help!

**Comment by sedabull** [comment]
@kevinsawicki, I have resolved this issue! Sortof... let me share what I know. First of all, the primary issue (me not being able to activate quantum-shell) turned out to be a simple mistake on my part. I made a stylistic decision to rename `lib/quantum-shell.coffee` to `lib/quantum-shell-controller.coffee`. However, I forgot to change the `"main"` field in my package.json, mostly because I made this change at the same time I updated to 0.178.0, and I incorrectly attributed the error to atom. This is still a bug though, IMO, because atom is apparently silently failing when a package's `"main"` field points to a non-exsistant file. I'd be happy to open a seperate issue for this if you think that's appropriate. Also, something I thought was a symptom of this error is in fact a seperate issue entirely. When you `apm link --dev` a package and then open atom normally, you are prompted to install the package(see the screenshot above) and trying to run `atom.packages.resolvePackagePath('my-package')` in the dev tools returns `undefined`. I have tested this on `quantum-shell` and the completely unrelated package `git-plus` on atom v-0.180.0. I'm resonably certain this is not the intended behavior, and I'd happily open an issue for this as well with everything I know.
